sdk/preview_termination_queries.go

```diff
@@ -0,0 +1,794 @@
+	sectorNumber uint64,
+	gasLimit uint64,
+	offchain bool,
+) (actor *types.ActorV5, totalBurn *corebig.Int, epoch abi.ChainEpoch, err error) {
+	lClient, closer, err := q.extern.ConnectLotusClient()
+	if err != nil {
+		return nil, nil, 0, err
+	}
+	defer closer()
+	api := *lClient
+
+	h, ts, err := parseTipSetAndHeight(ctx, api, tipset, vmHeight)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	// Lookup actor balance
+	actor, err = api.StateGetActor(ctx, minerAddr, ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	// Lookup current owner / worker
+	minerInfo, err := api.StateMinerInfo(ctx, minerAddr, ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	var params miner.TerminateSectorsParams
+
+	if gasLimit == 0 {
+		gasLimit = 6000000000
+	}
+
+	sectorPartition, err := api.StateSectorPartition(ctx, minerAddr, abi.SectorNumber(sectorNumber), ts.Key())
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	termination := miner.TerminationDeclaration{
+		Deadline:  sectorPartition.Deadline,
+		Partition: sectorPartition.Partition,
+		Sectors:   bitfield.NewFromSet([]uint64{sectorNumber}),
+	}
+	params = miner.TerminateSectorsParams{
+		Terminations: []miner.TerminationDeclaration{termination},
+	}
+
+	totalBurn, err = terminateSectors(ctx, *lClient, h, ts, minerAddr, minerInfo,
+		params, int64(gasLimit), offchain)
+	if err != nil {
+		return nil, nil, 0, err
+	}
+
+	return actor, totalBurn, h, nil
+}
+
+type terminationTask struct {
+	termination         miner.TerminationDeclaration
+	deadlineHeight      abi.ChainEpoch
+	deadlineTs          *types.TipSet
+	sectorsCount        int64
+	sampledSectorsCount int64
+}
+
+// PreviewTerminateSectors gets the burnt funds for when all the sectors are terminated
+func (q *fevmQueries) PreviewTerminateSectors(
+	ctx context.Context,
+	minerAddr address.Address,
+	tipset string,
+	vmHeight uint64,
+	batchSize uint64,
+	gasLimit uint64,
+	useSampling bool,
+	optimize bool,
+	offchain bool,
+	maxPartitions uint64,
+	errorCh chan error,
+	progressCh chan *poolstypes.PreviewTerminateSectorsProgress,
+	resultCh chan *poolstypes.PreviewTerminateSectorsReturn,
+) {
+	lClient, closer, err := q.extern.ConnectLotusClient()
+	if err != nil {
+		errorCh <- err
+		return
+	}
+	defer closer()
+	api := *lClient
+
+	h, ts, err := parseTipSetAndHeight(ctx, api, tipset, vmHeight)
+	if err != nil {
+		errorCh <- err
+		return
+	}
+
+	// Lookup actor balance
+	actor, err := api.StateGetActor(ctx, minerAddr, ts.Key())
+	if err != nil {
+		errorCh <- err
+		return
+	}
+
+	// Lookup current owner / worker
+	minerInfo, err := api.StateMinerInfo(ctx, minerAddr, ts.Key())
+	if err != nil {
+		errorCh <- err
+		return
+	}
+
+	workerActor, err := api.StateGetActor(ctx, minerInfo.Worker, ts.Key())
+	if err != nil {
+		errorCh <- err
+		return
+	}
+
+	provingDeadline, err := api.StateMinerProvingDeadline(ctx, minerAddr, ts.Key())
+	if err != nil {
+		errorCh <- err
+		return
+	}
+
+	prevHeight := h - 120
```

Comment:
why minus 120 here?

Reply:
It is impossible to terminate sectors during an immutable deadline ... at the time of the query, on a miner that is using all it's deadlines, some deadlines will be immutable, so I work around that by terminating the immutable deadline sectors using an earlier height where the deadline will be mutable. It's not perfect, but typically the numbers are going to be almost the same ... just off by small amounts of attoFIL.